--- client/third_party/google/oauth2/_client.py
+++ client/third_party/google/oauth2/_client.py
@@ -99,13 +99,22 @@ def _token_endpoint_request(request, token_uri, body):
     headers = {
         'content-type': _URLENCODED_CONTENT_TYPE,
     }
-
-    response = request(
-        method='POST', url=token_uri, headers=headers, body=body)
-
-    response_body = response.data.decode('utf-8')
-
-    if response.status != http_client.OK:
+    retry = 0
+    # retry to fetch token for maximum of two times if any internal failure
+    # occurs.
+    while retry < 2:
+      response = request(
+          method='POST', url=token_uri, headers=headers, body=body)
+
+      response_body = response.data.decode('utf-8')
+
+      if response.status == http_client.OK:
+        break
+      else:
+        error_desc = json.loads(response_body).get('error_description') or ''
+        if error_desc == 'internal_failure' and retry < 2:
+          retry += 1
+          continue
         _handle_error_response(response_body)
 
     response_data = json.loads(response_body)
